packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/internal/ModelsTreeIdsCache.ts

```diff
@@ -376,74 +355,194 @@ export class ModelsTreeIdsCache {
+    const result = new Map<ModelCategoryKey, number>();
+    if (input.length === 0) {
+      return result;
+    }
+
+    const reader = this._queryExecutor.createQueryReader(
+      {
+        ecsql: `
+          SELECT COUNT(*) elementsCount, e.Category.Id categoryId, e.Model.Id modelId
+          FROM ${this._hierarchyConfig.elementClassSpecification} e
+          WHERE
+            e.Parent.Id IS NULL AND (
+              ${input.map(({ modelId, categoryId }) => `e.Model.Id = ${modelId} AND e.Category.Id = ${categoryId}`).join(" OR ")}
+            )
+          GROUP BY e.Model.Id, e.Category.Id
+        `,
+      },
+      { rowFormat: "ECSqlPropertyNames", limit: "unbounded" },
+    );
+
+    for await (const row of reader) {
+      const key: ModelCategoryKey = `${row.modelId}-${row.categoryId}`;
+      result.set(key, row.elementsCount);
+    }
+    return result;
+  }
+
+  public async getRootCategoryElementsCount(modelId: ModelId, categoryId: CategoryId): Promise<number> {
+    return this._categoryElementCounts.getRootCategoryElementsCount(modelId, categoryId);
+  }
+
+  private async queryElementsChildrenInfo({ modelId }: { modelId: ModelId }): Promise<ModelParentMap> {
     const reader = this._queryExecutor.createQueryReader(
       {
         ctes: [
-          /* sql */ `
-            CategoryElements(id, modelId, categoryId) AS (
-              SELECT ECInstanceId, Model.Id, Category.Id
+          `
+          ElementsChildrenInfo(Id, CategoryId, ParentId, RootElementCategoryId, ParentHasParent) AS (
+              SELECT ECInstanceId, Category.Id, Parent.Id, Category.Id, false
               FROM ${this._hierarchyConfig.elementClassSpecification}
               WHERE
-                Parent.Id IS NULL
-                AND (
-                  ${input.map(({ modelId, categoryId }) => `Model.Id = ${modelId} AND Category.Id = ${categoryId}`).join(" OR ")}
-                )
+                Parent.Id IS NULL AND Model.Id = ${modelId}
 
               UNION ALL
 
-              SELECT c.ECInstanceId, p.modelId, p.categoryId
+              SELECT c.ECInstanceId, c.Category.Id, c.Parent.Id, p.RootElementCategoryId, IIF(p.ParentId IS NULL, false, true)
               FROM ${this._hierarchyConfig.elementClassSpecification} c
-              JOIN CategoryElements p ON c.Parent.Id = p.id
+              JOIN ElementsChildrenInfo p ON c.Parent.Id = p.Id
             )
           `,
         ],
         ecsql: `
-          SELECT modelId, categoryId, COUNT(id) elementsCount
-          FROM CategoryElements
-          GROUP BY modelId, categoryId
+          SELECT
+            this.CategoryId categoryId,
+            this.ParentId parentId,
+            this.Id id,
+            this.RootElementCategoryId rootElementCategoryId,
+            IFNULL((
+              SELECT 1
+              FROM (
+                SELECT ParentId FROM ElementsChildrenInfo
+              )
+              WHERE ParentId = this.Id
+              LIMIT 1
```

Comment:
why not just `SELECT 1 FROM ${this._hierarchyConfig.elementClassSpecification} WHERE Parent.Id = this.Id`?